pageserver/src/http/routes.rs

```diff
@@ -1423,6 +1425,76 @@ async fn list_location_config_handler(
     json_response(StatusCode::OK, result)
 }
 
+// Do a time travel recovery on the given tenant/tenant shard. Tenant needs to be detached
+// (from all pageservers) as it invalidates consistency assumptions.
+async fn tenant_time_travel_remote_storage_handler(
+    request: Request<Body>,
+    cancel: CancellationToken,
+) -> Result<Response<Body>, ApiError> {
+    let tenant_shard_id: TenantShardId = parse_request_param(&request, "tenant_shard_id")?;
+
+    check_permission(&request, Some(tenant_shard_id.tenant_id))?;
+
+    let timestamp_raw = must_get_query_param(&request, "travel_to")?;
+    let timestamp = humantime::parse_rfc3339(&timestamp_raw)
+        .with_context(|| format!("Invalid time for travel_to: {timestamp_raw:?}"))
+        .map_err(ApiError::BadRequest)?;
+
+    let done_if_after_raw = must_get_query_param(&request, "done_if_after")?;
+    let done_if_after = humantime::parse_rfc3339(&done_if_after_raw)
+        .with_context(|| format!("Invalid time for done_if_after: {done_if_after_raw:?}"))
+        .map_err(ApiError::BadRequest)?;
+
+    // This is just a sanity check to fend off naive wrong usages of the API:
+    // the tenant needs to be detached *everywhere*
+    let state = get_state(&request);
+    let we_manage_tenant = state.tenant_manager.manages_tenant_shard(tenant_shard_id);
+    if we_manage_tenant {
+        return Err(ApiError::BadRequest(anyhow!(
+            "Tenant {tenant_shard_id} is already attached at this pageserver"
+        )));
+    }
+
+    let Some(storage) = state.remote_storage.as_ref() else {
+        return Err(ApiError::InternalServerError(anyhow::anyhow!(
+            "remote storage not configured, cannot run time travel"
+        )));
+    };
+
+    if timestamp > done_if_after {
+        return Err(ApiError::BadRequest(anyhow!(
+            "The done_if_after timestamp comes before the timestamp to recover to"
+        )));
+    }
+
+    tracing::info!("Issuing time travel request internally. timestamp={timestamp_raw}, done_if_after={timestamp_raw}");
+
+    remote_timeline_client::upload::time_travel_recover_tenant(
+        storage,
+        &tenant_shard_id,
+        timestamp,
+        done_if_after,
+        &cancel,
```

Comment:
Does it make sense to pass in the connection cancellation token? I think we have backoff retries at different levels, so the whole operation restarted few times could take a while.

Should we guard against pagerserver attaching while this is happening?

Reply:
> Does it make sense to pass in the connection cancellation token? 

I think it does, we really don't want the time travel recovery to be ongoing when we attach it to a pageserver. it must be cancelled at that point. Think of the following scenario:

1. control plane sends request to PS A, it starts uncancellable background process
2. control plane sends request to PS A again, it recognizes that the process is still ongoing
3. there is network loss between control plane and PS A, control plane's third retry therefore fails
4. control plane chooses PS B instead and sends its next request there 
5. request finishes at PS B
6. control plane attaches PS B and PS receives data, writing it to remote storage.
7. meanwhile, PS A had some networking issues but now it has network again. it finishes the background process from 1 and writes a partially old state to remote storage

that's I think what generations are meant to solve but I don't think we can use generations here.

> the whole operation restarted few times could take a while.

definitely, it can take minutes. I think we should ask control plane to keep open the TCP connection and HTTP stream for a long time (think 30 minutes), and only retry if it gets closed or times out either due to TCP timeout.

If the timeout of the pageserver is known to the control plane, then it can derive from that when it can calculate when the connection will have timed out on the pageserver's side as well.

Reply:
Hmmm thinking about it more, this is precisely what generations are meant to prevent. If control plane, in step 6, makes a new generation, uploads for the old generation in step 7 will be ignored. So I think we should be fine.

Still, ideally this would already be prevented with pageserver internal means.

Reply:
> Well, these can be of course follow-ups.

I'd like to make them follow-ups, yes.

Reply:
I'll leave that to you then, good to resolve.